cs/src/core/Index/FasterLog/FasterLogIterator.cs

```diff
@@ -484,11 +513,10 @@ private unsafe bool GetNextInternal(out long physicalAddress, out int entryLengt
                 if (Utility.MonotonicUpdate(ref NextAddress, currentAddress, out long oldCurrentAddress))
                 {
                     currentAddress = oldCurrentAddress;
+                    nextAddress = NextAddress;
```

Comment:
Not sure if this code can be run by multiple threads simultaneously, but if it can then maybe: 
`nextAddress = currentAddress;`

Reply:
Ah yes, you are right.

Otherwise it defeats the purpose of adding this extra parameter, since one could just check NextAddress as you mentioned in your message earlier.

Reply:
Just so everyone gets the context, the situation this catches is that there are two GetNext calls in parallel. The caller can safely finalize only until the immediate next address _for their own GetNext_, because from their perspective they have completed the processing only of the row they just retrieved.

Reply:
This above argument does assume that when a caller finishes processing a row, they somehow know that all previous rows have also completed processing, so it's safe to finalize until the end of that address.

Reply:
See here for a discussion: https://github.com/microsoft/FASTER/issues/244#issuecomment-593519144

Reply:
Good catch. I will update the PR as suggested.
Re: the "completed up to" situation, the reading entity probably needs to hold on to both the current and the next addresses in some synchronized collection(s), and only issue a CompleteUntil() up to the point all known nexts are accounted for by the known currents.